python/lsst/daf/butler/core/fileTemplates.py

```diff
@@ -0,0 +1,137 @@
+    ----------
+    template : `str`
+        Template string.
+
+    Notes
+    -----
+    The templates use the standard Format Specification Mini-Language
+    with the caveat that only named fields can be used. The field names
+    are taken from the DataUnits along with two additional fields:
+    "datasettype" will be replaced with the DataSetType and "component"
+    will be replaced with the component name of a composite.
+
+    The mini-language is extended to understand a "?" in the format
+    specification. This indicates that a field is optional. If that
+    DataUnit is missing the field, along with the text before the field,
+    unless it is a path separator, will be removed from the output path.
+    """
+
+    def __init__(self, template):
+        self.template = template
+
+    def format(self, dataunits, datasettype=None, component=None):
+        """Format a template string into a full path.
+
+        Parameters
+        ----------
+        dataunits : `DataUnits`
+            DataUnits and the corresponding values.
+        datasettype : `str`, optional.
+            DataSetType name to use if needed. If it contains a "." separator
+            the type name will be split up into the main DataSetType and a
+            component.
+        component : `str`, optional
+            Component of a composite. If `datasettype` defines a component
+            this parameter will be ignored.
+
+        Returns
+        -------
+        path : `str`
+            Expanded path.
+
+        Raises
+        ------
+        KeyError
+            Requested field is not defined and the field is not optional.
+            Or, `component` is specified but "component" was not part of
+            the template.
+        """
+        units = dataunits.definedUnits()
+
+        if datasettype is not None:
+            # calexp.wcs means wcs component of a calexp
+            if "." in datasettype:
+                datasettype, component = datasettype.split(".", maxsplit=1)
+            units["datasettype"] = datasettype
+
+        usedComponent = False
+        if component is not None:
+            units["component"] = component
+
+        fmt = string.Formatter()
+        parts = fmt.parse(self.template)
+        output = ""
+
+        for literal, field_name, format_spec, conversion in parts:
+
+            if field_name == "component":
+                usedComponent = True
+
+            if "?" in format_spec:
+                optional = True
+                # Remove the non-standard character from the spec
+                format_spec = format_spec.replace("?", "")
+            else:
+                optional = False
+
+            if field_name in units:
+                value = units[field_name]
+            elif optional:
+                # If this is optional ignore the format spec
+                # and do not include the literal text prior to the optional
+                # field unless it contains a "/" path separator
```

Comment:
Will it now have two path separators? E.g. `/{visit}/{hello:?}/{camera} -> /3//hsc`.

Reply:
no because of the path cleanup on the last line of the method.